### PR TITLE
Enabling psalm on php >= 7.1, fixing #125, #168, #169, partially fixing #177

### LINKED ISSUE #177: Using polyfill for ext-intl if
As I stated before in #175 it could be problematic to just disable functionality when the extension is not available on the system. (We had the problem after upgrading from 2.1.4 to 2.1.5.)

Using a polyfill like https://github.com/symfony/polyfill-intl-icu could make this package work out of the box and still use the original functions from ext-intl if available. The suggestion for ext-intl in the composer.json should of course be kept.

The question is wether or not the polyfill from above or another one would make the code work.

For example the package symfony/intl does the same, see https://github.com/symfony/intl/blob/master/composer.json

### PR DESCRIPTION
**Pull request by SignpostMarv**
* re: 210ac3c & da2e0c6, there's an `@psalm-suppress` or two in there rather than a baseline file because the older versions of psalm required for `php < 7` consider the baseline attribute invalid so that'd be best put off until you make the decision to drop support for php 5 in a given release.
* c504af2 only partially resolves #177 because the `SpoofChecker` class isn't polyfilled. The options there would be to find a polyfill just for SpoofChecker, or to fork out the `SpoofCheckValidation` class into a package that explicitly requires `ext-intl:*`
* 6812a28 & 02a6193 were done for value-added giggles.

### CODE REVIEW

**Comment by egulias** [comment]
Hi @SignpostMarv great PR, thanks. I'll take a deeper look into Psalm. However, I'm relying on [Scrutinizer](https://scrutinizer-ci.com/g/egulias/EmailValidator/) for static analysis and would like to know if it is worth to have both.

**Comment by SignpostMarv** [comment]
I've been using a combination of phpstan, psalm, and scrutinizer for quite a while now- my experience is that scrutinizer doesn't catch everything and that running static analysis on your local machine leads to a more rapid development cycle as you're not waiting for CI to spin up multiple instances & finish all the matrices. This leaves scrutinizer's main purpose down to tracking overall code quality & code coverage.

sebastian/phpcpd can be useful in identifying deduplication issues before they reach scrutinizer.

povils/phpmnd can help identify integers that should probably be consts.

maglnet/composer-require-checker checks if your code uses any php extensions or downstream code not explicitly required by the project.

**Comment by k00ni** [comment]
When do you plan to merge this?

File: EmailValidator/Validation/DNSCheckValidation.php [hunk 1]
```diff
@@ -15,13 +15,13 @@ class DNSCheckValidation implements EmailValidation
...
+     * @var InvalidEmail|null
      */
     private $error;
-    
+
     public function __construct()
     {
-        if (!extension_loaded('intl')) {
+        if (!function_exists('idn_to_ascii')) {
             throw new \LogicException(sprintf('The %s class requires the Intl extension.', __CLASS__));
```
Comment (egulias) [inline EmailValidator/Validation/DNSCheckValidation.php]: How does `idn_to_ascii` correlates to `intl` extension? Are they the same?
If so, please also update the error message.

File: composer.json [hunk 2]
```diff
@@ -13,19 +13,15 @@
       "dev-master": "2.1.x-dev"
     }
   },
-  "repositories": [
-    {
-      "type": "git",
-      "url": "https://github.com/dominicsayers/isemail"
```
Comment (egulias) [inline composer.json]: Why are you removing this? it is needed for the tests comparision :)

File: EmailValidator/Validation/DNSCheckValidation.php (hunk 1 above)
Comment (SignpostMarv) [reply EmailValidator/Validation/DNSCheckValidation.php]: `symfony/polyfill-intl-idn` will take over when the extension isn't loaded, thus the condition needs to check for the function used.

File: composer.json (hunk 2 above)
Comment (SignpostMarv) [reply composer.json]: as mentioned in the commit message, the package is listed on packagist & directly listing the repo causes install issues on windows & git bash.

**Comment by GrahamCampbell** [comment]
This PR does a lot of different things, which makes it hard to merge. I think it should be broken up into more atomic pieces. Not all of the changes are still relevant now too.

**Comment by SignpostMarv** [comment]
checking a rebase builds before splitting things off.

**Comment by egulias** [comment]
@SignpostMarv apparently Psalm version you are requiring does not supports php 7.0 (https://travis-ci.org/egulias/EmailValidator/jobs/631128083?utm_medium=notification&utm_source=github_status)

**Comment by SignpostMarv** [comment]
> @SignpostMarv apparently Psalm version you are requiring does not supports php 7.0 (https://travis-ci.org/egulias/EmailValidator/jobs/631128083?utm_medium=notification&utm_source=github_status)

no version constraint is applied, so this may occasionally fail (necessitating an adjustment to the matrix as in 079ac0e)

File: composer.json [hunk 3]
```diff
@@ -14,9 +14,11 @@
   },
   "require": {
     "php": ">=5.5",
-    "doctrine/lexer": "^1.0.1"
+    "doctrine/lexer": "^1.0.1",
+    "symfony/polyfill-intl-idn": "^1.10"
   },
   "require-dev": {
+    "ext-intl": "*",
```
Comment (egulias) [inline composer.json]: Relates to #231 , since intl is still required for the spoof checker.
Maybe you can revert this change to avoid blocking this PR because of the INTL dependency fix.

Comment (SignpostMarv) [reply composer.json]: re: `SpoofChecker`, what're your thoughts [regarding forking `SpoofCheckValidation` to it's own package that requires `ext-intl`](https://github.com/egulias/EmailValidator/pull/229#discussion_r362545036) ?

Comment (egulias) [reply composer.json]: Answered. However, we can still remove this changes from the current PR in order to merge it and solve the INTL issue there.

Comment (SignpostMarv) [reply composer.json]: removed in affd3dd
